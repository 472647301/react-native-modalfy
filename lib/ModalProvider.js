/* @flow */

import React, { Fragment, Component } from 'react'
import { BackHandler, View } from 'react-native'

import type {
  EventName,
  EventSubscription,
  ModalEventListeners,
  ModalName,
  Stack,
  StackItem,
} from '../types'
import { invariant, validateListener } from '../utils'
import ModalContext from './ModalContext'
import ModalStack from './ModalStack'

type Props = {
  children: React$Element<*>,
  stack: Stack<Set<StackItem>>,
}

type State = {
  currentModal: ?ModalName,
  closeAllModals: () => void,
  closeModal: (stackItem: StackItem) => void,
  closeModals: (modalName: ModalName) => void,
  getParams: (hash: ?string, fallback?: any) => any,
  openModal: (modalName: ModalName, params?: Object) => void,
  stack: Stack<Set<StackItem>>,
}

class ModalProvider extends Component<Props, State> {
  constructor(props: Props) {
    super(props)

    this.state = {
      currentModal: null,
      stack: props.stack,
      openModal: this._openModal,
      closeAllModals: this._closeAllModals,
      closeModal: this._closeModal,
      closeModals: this._closeModals,
      getParams: this._getParams,
    }
  }

  static displayName = 'ModalProvider'

  modalEventListeners: ModalEventListeners = new Set()

  componentDidMount() {
    const { stack } = this.props
    invariant(stack, 'You need to provide a stack prop to <ModalProvider>')
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this._handleBackPress)
  }

  _clearListeners = (hash: string) => {
    this.modalEventListeners.forEach(item => {
      if (item.event.includes(hash)) this.modalEventListeners.delete(item)
    })
  }

  _openModal = (modalName: ModalName, params: ?Object) => {
    const { content, names } = this.state.stack

    invariant(modalName?.length > 0, "You didn't pass any modal name")
    invariant(
      names.some(name => name === modalName),
      // $FlowFixMe
      `'${modalName}' is not a valid modal name. Did you mean any of these: ${names.map(
        validName => `\n• ${validName}`
      )}`
    )

    const stackItem = content.find(item => item.name === modalName)
    const hash = `${modalName}_${Math.random()
      .toString(36)
      .substr(2, 9)}`

    if (!this.state.currentModal) {
      BackHandler.addEventListener('hardwareBackPress', this._handleBackPress)
    }

    this.setState(state => ({
      currentModal: modalName,
      stack: {
        ...state.stack,
        openedItems: state.stack.openedItems.add({
          ...stackItem,
          hash,
          ...(params && { params }),
        }),
      },
    }))
  }

  _closeAllModals = () => {
    const { openedItems } = this.state.stack

    openedItems.clear()

    this.setState(state => ({
      currentModal: null,
      stack: { ...state.stack, openedItems },
    }))
  }

  _closeModal = (stackItem: StackItem) => {
    const { openedItems } = this.state.stack

    openedItems.delete(stackItem)

    const openedItemsArray = Array.from(openedItems)

    this.setState(state => ({
      currentModal: openedItemsArray?.[openedItemsArray?.length - 1]?.name,
      stack: { ...state.stack, openedItems },
    }))
  }

  _closeModals = (modalName: ModalName) => {
    const { openedItems: oldOpenedItems, names } = this.state.stack

    invariant(
      modalName?.length > 0,
      "You didn't pass any modal name to closeModals()"
    )
    invariant(
      names.some(name => name === modalName),
      // $FlowFixMe
      `'${modalName}' is not a valid modal name. Did you mean any of these: ${names.map(
        validName => `\n• ${validName}`
      )}`
    )

    const newOpenedItems = new Set(oldOpenedItems)

    newOpenedItems.forEach(item => {
      if (item.name === modalName) newOpenedItems.delete(item)
    })

    if (newOpenedItems.size !== oldOpenedItems.size) {
      const openedItemsArray = Array.from(newOpenedItems)
      this.setState(state => ({
        currentModal: openedItemsArray?.[openedItemsArray?.length - 1]?.name,
        stack: { ...state.stack, openedItems: newOpenedItems },
      }))
    }
  }

  _getParams = (hash: ?string, fallback?: any): any => {
    const { openedItems } = this.state.stack
    let stackItem

    openedItems.forEach(item => {
      if (item.hash === hash) stackItem = item
    })

    return stackItem?.params || fallback
  }

  _handleBackPress = () => {
    const {
      defaultOptions: { backButtonBehavior },
      openedItems,
    } = this.state.stack
    const { currentModal } = this.state

    if (currentModal) {
      if (backButtonBehavior === 'none') return true
      else if (backButtonBehavior === 'clear') {
        this.setState(state => ({
          stack: { ...state.stack, openedItems: new Set() },
        }))
        return true
      } else if (backButtonBehavior === 'pop') {
        const openedItemsArray = [...openedItems]
        this._closeModal(openedItemsArray.slice(-1)[0])
        return true
      }
    }

    return false
  }

  _registerListener = (
    hash: string,
    eventName: EventName,
    handler: () => void
  ): EventSubscription => {
    validateListener('add', { eventName, handler })
    const newListener = {
      event: `${hash}_${eventName}`,
      handler,
    }

    this.modalEventListeners.add(newListener)

    return {
      remove: () => this.modalEventListeners.delete(newListener),
    }
  }

  render() {
    const { children } = this.props
    return (
      <ModalContext.Provider value={this.state}>
        <Fragment>
          <View style={{ flex: 1 }}>{children}</View>
          <ModalStack
            {...this.state}
            eventListeners={this.modalEventListeners}
            registerListener={this._registerListener}
            clearListeners={this._clearListeners}
          />
        </Fragment>
      </ModalContext.Provider>
    )
  }
}

export default ModalProvider
