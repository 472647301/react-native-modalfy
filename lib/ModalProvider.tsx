import React, { ReactNode, useEffect, useRef, useState } from 'react'
import { BackHandler } from 'react-native'
import { useCallback } from 'use-memo-one'

import {
  ModalStack as ModalStackType,
  ModalStateSubscription,
  ModalContextProvider,
  ModalEventListeners,
  ModalStateListener,
  SharedProps,
} from '../types'

import ModalContext from './ModalContext'
import ModalStack from './ModalStack'
import ModalState from './ModalState'

import { invariant, validateListener } from '../utils'

interface Props {
  children: ReactNode
  stack: ModalStackType<any>
}

/**
 * `<ModalProvider>` is the component you're going to use to wrap your whole application,
 * so it'll be able to display your modals on top of everything else, using React Context API.
 *
 * @prop { ModalStackType } `stack` - Modal stack object generated by `createModalStack()`
 *
 * @see https://colorfy-software.gitbook.io/react-native-modalfy/guides/stack#provider
 */
const ModalProvider = ({ children, stack }: Props) => {
  const modalStateSubscription =
    useRef<ModalStateSubscription<any> | undefined>()

  const modalEventListeners = useRef<ModalEventListeners>(new Set()).current

  const openModal: SharedProps<any>['openModal'] = (modalName, params) => {
    const { currentModal } = ModalState.getState()

    if (!currentModal) {
      BackHandler.addEventListener(
        'hardwareBackPress',
        ModalState.handleBackPress,
      )
    }

    ModalState.openModal(modalName, params)
  }

  const getParam: SharedProps<any>['getParam'] = (
    hash,
    paramName,
    defaultValue,
  ) => ModalState.getParam(hash, paramName, defaultValue)

  const closeModal: SharedProps<any>['closeModal'] = (stackItem) =>
    ModalState.closeModal(stackItem)

  const closeModals: SharedProps<any>['closeModals'] = (modalName) =>
    ModalState.closeModals(modalName)

  const closeAllModals: SharedProps<any>['closeAllModals'] = () =>
    ModalState.closeAllModals()

  const [contextValue, setContextValue] = useState<
    ModalContextProvider<any, any>
  >({
    currentModal: null,
    closeAllModals,
    closeModals,
    closeModal,
    openModal,
    getParam,
    stack,
  })

  const registerListener: SharedProps<any>['registerListener'] = useCallback(
    (hash, eventName, handler) => {
      validateListener('add', { eventName, handler })
      const newListener = {
        event: `${hash}_${eventName}`,
        handler,
      }

      modalEventListeners.add(newListener)

      return {
        remove: () => modalEventListeners.delete(newListener),
      }
    },
    [modalEventListeners],
  )

  const clearListeners: SharedProps<any>['clearListeners'] = useCallback(
    (hash) => {
      modalEventListeners.forEach((item) => {
        if (item.event.includes(hash)) modalEventListeners.delete(item)
      })
    },
    [modalEventListeners],
  )

  const listener: ModalStateListener<any> = (modalState, error) => {
    if (modalState) {
      setContextValue({
        ...contextValue,
        currentModal: modalState.currentModal,
        stack: modalState.stack,
      })
    } else console.warn('Modalfy', error)
  }

  useEffect(() => {
    invariant(stack, 'You need to provide a `stack` prop to <ModalProvider>')

    ModalState.init<any>(() => ({
      currentModal: null,
      stack,
    }))

    modalStateSubscription.current = ModalState.subscribe(listener)

    return () => {
      BackHandler.removeEventListener(
        'hardwareBackPress',
        ModalState.handleBackPress,
      )
      modalStateSubscription.current?.unsubscribe()
    }

    // NOTE: Should only be triggered on initial mount and return when unmounted
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <ModalContext.Provider value={contextValue}>
      <>
        {children}
        <ModalStack
          {...contextValue}
          clearListeners={clearListeners}
          registerListener={registerListener}
          eventListeners={modalEventListeners}
          removeClosingAction={ModalState.removeClosingAction}
        />
      </>
    </ModalContext.Provider>
  )
}

export default ModalProvider
