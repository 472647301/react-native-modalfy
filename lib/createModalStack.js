/* @flow */

import type { Config, Options, Stack } from '../types'
import {
  invariant,
  getStackItemData,
  defaultOptions,
  validateDefaultOptions,
} from '../utils'

export default function(config: Config, customDefaultOptions: Options): Stack {
  invariant(config, 'You need to provide a config to createModalStack()')
  validateDefaultOptions(customDefaultOptions)

  const initialStack: Stack = {
    names: [],
    content: [],
    defaultOptions: {
      ...defaultOptions,
      ...customDefaultOptions,
    },
    openedItems: [],
    total: 0,
  }

  return Object.entries(config).reduce((output, entry, index) => {
    const { name, component, options } = getStackItemData(entry[0], entry[1])
    return {
      ...output,
      names: [...output.names, name],
      content: [
        ...output.content,
        { index, name, component, ...(options && { options }) },
      ],
      total: output.total + 1,
    }
  }, initialStack)
}
