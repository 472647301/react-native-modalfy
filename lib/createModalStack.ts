import { ModalStack, ModalStackConfig, ModalOptions } from '../types'

import {
  validateDefaultOptions,
  getStackItemData,
  defaultOptions,
  invariant,
} from '../utils'

/**
 * `createModalStack()` is the function that's going to turn your configuration into a usable modal stack.
 *
 * @argument { ModalStackConfig } config - Modal stack configuration.
 * @argument { ModalOptions } [customDefaultOptions] - Configuration options to apply to all modals by default (optional).
 *
 * @returns { ModalStack } Modal stack configuration object to provide to `<ModalProvider>`'s `stack` prop.
 *
 * @see https://colorfy-software.gitbook.io/react-native-modalfy/guides/stack
 */
export default function <P>(
  config: ModalStackConfig,
  customDefaultOptions?: ModalOptions,
): ModalStack<P> {
  invariant(config, 'You need to provide a config to createModalStack()')
  validateDefaultOptions(customDefaultOptions)

  const initialStack: ModalStack<P> = {
    names: [],
    content: [],
    defaultOptions: {
      ...defaultOptions,
      ...customDefaultOptions,
    },
    openedItems: new Set(),
    openedItemsSize: 0,
  }

  // @ts-ignore
  return Object.entries(config).reduce<ModalStack<P>>(
    // @ts-ignore
    (output, entry, index) => {
      const { name, component, options } = getStackItemData(entry[0], entry[1])
      return {
        ...output,
        names: [...output.names, name],
        content: [
          ...output.content,
          { index, name, component, hash: '', ...(options && { options }) },
        ],
      }
    },
    initialStack,
  )
}
