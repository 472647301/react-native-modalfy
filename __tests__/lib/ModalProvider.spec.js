import React, { Fragment } from 'react'
import TestRenderer from 'react-test-renderer'
import { View } from 'react-native'
import { shallow } from 'enzyme'

import ModalProvider from '../../lib/ModalProvider'
import createModalStack from '../../lib/createModalStack'

describe('ðŸ’§ ModalProvider', () => {
  beforeAll(() => {
    stack = createModalStack({ Modal: View })
  })

  test('should throw when no stack was passed', () => {
    expect(() => shallow(<ModalProvider />)).toThrow(
      'You need to provide a stack prop to <ModalProvider>'
    )
  })

  test('should render children correctly', () => {
    const component = TestRenderer.create(
      <ModalProvider stack={stack}>
        <View />
      </ModalProvider>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
