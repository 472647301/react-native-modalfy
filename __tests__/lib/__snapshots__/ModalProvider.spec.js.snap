// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ’§ ModalProvider should render children correctly 1`] = `
Array [
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View />
  </View>,
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 1334,
          },
        ],
        "zIndex": 0,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "black",
          "height": 1334,
          "opacity": 0,
          "width": 750,
        }
      }
    />
  </View>,
]
`;
